internal/k8s/helpers.go

```diff
@@ -83,3 +89,77 @@ func isStatusEqual(objA, objB interface{}) bool {
 	}
 	return false
 }
+
+// IsObjectEqual checks if objects received during update are equal.
+//
+// Make an attempt to avoid comparing full objects since it can be very CPU intensive.
+// Prefer comparing Generation when only interested in spec changes.
+func IsObjectEqual(old, new interface{}) (bool, error) {
+
+	// Fast path for any object: when ResourceVersions are equal, the objects are equal.
+	// NOTE: This optimizes the case when controller-runtime executes full sync and sends updates for all objects.
+	if isResourceVersionEqual(old, new) {
+		return true, nil
+	}
+
+	switch old := old.(type) {
+
+	// Fast path for objects that implement Generation and where only spec changes matter.
+	// Generation is implemented in CRDs, Ingress and IngressClass.
+	case *contour_api_v1alpha1.ExtensionService:
+		return isGenerationEqual(old, new), nil
+	case *contour_api_v1.TLSCertificateDelegation:
+		return isGenerationEqual(old, new), nil
+	case *contour_api_v1alpha1.ContourConfiguration:
+		return isGenerationEqual(old, new), nil
+
+	case *gatewayapi_v1beta1.GatewayClass:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1beta1.Gateway:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1beta1.HTTPRoute:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1alpha2.TLSRoute:
+		return isGenerationEqual(old, new), nil
+
+	// Slow path: compare the content of the objects.
+	case *contour_api_v1.HTTPProxy:
+		if new, ok := new.(*contour_api_v1.HTTPProxy); ok {
+			return isGenerationEqual(old, new) &&
+				apiequality.Semantic.DeepEqual(old.ObjectMeta.Annotations, new.ObjectMeta.Annotations), nil
```

Comment:
I _think_ this is reasonable comparison logic, though worth noting that previously, Label changes (for example) would trigger a DAG rebuild whereas now they won't. I can't think of a reason that they would need to trigger a rebuild, but it is a slight change in behavior. Similarly for some of the below objects.

Reply:
I can add more fields, but it seems that only annotations would get used by Contour. And e.g. object name is immutable, so it is safe to exclude.

Adding everything is bit problematic since `DeepEqual()` and alikes work in opposite way to `cmp.Equal()` which gave the option to ignore named fields.

Reply:
I think it's OK as is; just need to squint at each individual object to make sure we're comparing the full set of things.